lib/web_ui/dev/roll_fallback_fonts.dart

```diff
@@ -509,3 +494,364 @@ bool _checkForLicenseAttribution(Uint8List fontBytes) {
   }
   return false;
 }
+
+class _Font {
+  _Font(this.family, this.index, this.starts, this.ends);
+
+  final String family;
+  final int index;
+  final List<int> starts;
+  final List<int> ends; // inclusive ends
+
+  static int compare(_Font a, _Font b) => a.index.compareTo(b.index);
+
+  String get shortName =>
+      _shortName +
+      String.fromCharCodes(
+          '$index'.codeUnits.map((int ch) => ch - 48 + 0x2080));
+
+  String get _shortName => family.startsWith('Noto Sans ')
+      ? family.substring('Noto Sans '.length)
+      : family;
+}
+
+/// The boundary of a range of a font.
+class _Boundary {
+  _Boundary(this.value, this.isStart, this.font);
+  final int value; // inclusive start or exclusive end.
+  final bool isStart;
+  final _Font font;
+
+  static int compare(_Boundary a, _Boundary b) => a.value.compareTo(b.value);
+}
+
+class _Range {
+  _Range(this.start, this.end, this.fontSet);
+  final int start;
+  final int end;
+  final _FontSet fontSet;
+
+  @override
+  String toString() {
+    return '[${start.toRadixString(16)}, ${end.toRadixString(16)}]'
+        ' (${end - start + 1})'
+        ' ${fontSet.description()}';
+  }
+}
+
+/// A canonical representative for a set of _Fonts. The fonts are stored in
+/// order of increasing `_Font.index`.
+class _FontSet {
+  _FontSet(this.fonts);
+
+  /// The number of [_Font]s in this set.
+  int get length => fonts.length;
+
+  /// The members of this set.
+  final List<_Font> fonts;
+
+  /// Number of unicode ranges that are supported by this set of fonts.
+  int rangeCount = 0;
+
+  /// The serialization order of this set. This index is assigned after building
+  /// all the sets.
+  late final int index;
+
+  static int orderByDecreasingRangeCount(_FontSet a, _FontSet b) {
+    final int r = b.rangeCount.compareTo(a.rangeCount);
+    if (r != 0) {
+      return r;
+    }
+    return orderByLexicographicFontIndexes(a, b);
+  }
+
+  static int orderByLexicographicFontIndexes(_FontSet a, _FontSet b) {
+    for (int i = 0; i < a.length && i < b.length; i++) {
+      final int r = _Font.compare(a.fonts[i], b.fonts[i]);
+      if (r != 0) {
+        return r;
+      }
+    }
+    assert(a.length != b.length); // _FontSets are canonical.
+    return a.length - b.length;
+  }
+
+  @override
+  String toString() {
+    return description();
+  }
+
+  String description() {
+    return fonts.map((_Font font) => font.shortName).join(', ');
+  }
+}
+
+/// A trie node [1] used to find the canonical _FontSet.
+///
+/// [1]: https://en.wikipedia.org/wiki/Trie
+class _TrieNode {
```

Comment:
Yes, this is somewhat vestigial of a more complicated scheme I had.

We could use a set of FontSets, but it would require a custom hashCode and equality, which is more code than the trie. (We would probably want a Map, rather than a Set so we have a distinct representative object for each set that has the index as a field.)

---

Before I simplified away a couple of pages of code, the parent structure of the trie was useful in differentially encoding one FontSet against the previous. We could save another 1kB of compiled JavaScript size with the more complicated scheme, but we also have to maintain the code. There are simpler ways of getting a similar saving (not implemented here), and the differential encoding _requires_ all the fontsets are decoded together. We don't decode them on need, but we could shave a few ms (<4ms)  from first-frame-with-uncovered-codepoints by decoding on demand.